### ISSUE TITLE
invalid syntax in conditional- Kernel 5.10.14 on OpenSUSE 15.2 with AOCC

### ISSUE THREAD

**Issue by michaelfuckner** [ClangBuiltLinux/linux#1298]
I am trying to compile on openSUSE 15.2 with AMDs AOCC. Fresh kernel sources, make menuconfig, accept and save all defaults and make LLVM=1. Looks like there is something easy I am missing, but can you help me which point I am missing?

Thanks,
 Michael!

ram1:~ # source /opt/AMD/setenv_AOCC.sh 
ram1:~ # clang -v
AMD clang version 11.0.0 (CLANG: AOCC_2.3.0-Build#85 2020_11_10) (based on LLVM Mirror.Version.11.0.0)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/AMD/aocc-compiler-2.3.0/bin
Found candidate GCC installation: /usr/lib64/gcc/x86_64-suse-linux/10
Found candidate GCC installation: /usr/lib64/gcc/x86_64-suse-linux/7
Found candidate GCC installation: /usr/lib64/gcc/x86_64-suse-linux/9
Selected GCC installation: /usr/lib64/gcc/x86_64-suse-linux/10
Candidate multilib: .;@m64
Selected multilib: .;@m64
ram1:~ # cd /usr/src/linux-5.10.14/
ram1:/usr/src/linux-5.10.14 # make LLVM=1
include/config/auto.conf.cmd:1369: *** invalid syntax in conditional.  Stop.
ram1:/usr/src/linux-5.10.14 #

**Comment by nathanchance** [comment]
`make` really does not like the `#` in the version string :/

```
$ sed -n 1369p build/x86_64/include/config/auto.conf.cmd
ifneq "$(CC_VERSION_TEXT)" "AMD clang version 11.0.0 (CLANG: AOCC_2.3.0-Build#85 2020_11_10) (based on LLVM Mirror.Version.11.0.0)"
```

This gets it working for me (although I run into a bunch of errors building `objtool` but that is tangential to this). @masahir0y would this be a proper fix?

```diff
diff --git a/Makefile b/Makefile
index b62d2d4ea7b0..e9d862ca5868 100644
--- a/Makefile
+++ b/Makefile
@@ -582,7 +582,7 @@ endif
 # Some architectures define CROSS_COMPILE in arch/$(SRCARCH)/Makefile.
 # CC_VERSION_TEXT is referenced from Kconfig (so it needs export),
 # and from include/config/auto.conf.cmd to detect the compiler upgrade.
-CC_VERSION_TEXT = $(shell $(CC) --version 2>/dev/null | head -n 1)
+CC_VERSION_TEXT = $(shell $(CC) --version 2>/dev/null | head -n 1 | sed 's/\#//')

 ifdef config-build
 # ===========================================================================
```

**Comment by nathanchance** [comment]
```diff
From 1a605d60b5d509f473abb10c581c61f527c7f1de Mon Sep 17 00:00:00 2001
From: Nathan Chancellor <nathan@kernel.org>
Date: Tue, 16 Feb 2021 12:21:44 -0700
Subject: [PATCH] Makefile: Remove # characters from compiler string

When using AMD's Optimizing C/C++ Compiler (AOCC), the build fails due
to a # character in the version string, which is interpreted as a
comment:

$ make CC=clang defconfig init/main.o
include/config/auto.conf.cmd:1374: *** invalid syntax in conditional. Stop.

$ sed -n 1374p include/config/auto.conf.cmd
ifneq "$(CC_VERSION_TEXT)" "AMD clang version 11.0.0 (CLANG: AOCC_2.3.0-Build#85 2020_11_10) (based on LLVM Mirror.Version.11.0.0)"

Remove all # characters in the version string so that the build does not
fail unexpectedly.

Link: https://github.com/ClangBuiltLinux/linux/issues/1298
Signed-off-by: Nathan Chancellor <nathan@kernel.org>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index de1acaefe87e..9bfe1307447e 100644
--- a/Makefile
+++ b/Makefile
@@ -581,7 +581,7 @@ endif
 # Some architectures define CROSS_COMPILE in arch/$(SRCARCH)/Makefile.
 # CC_VERSION_TEXT is referenced from Kconfig (so it needs export),
 # and from include/config/auto.conf.cmd to detect the compiler upgrade.
-CC_VERSION_TEXT = $(shell $(CC) --version 2>/dev/null | head -n 1)
+CC_VERSION_TEXT = $(shell $(CC) --version 2>/dev/null | head -n 1 | sed 's/\#//g')
 
 ifdef config-build
 # ===========================================================================
-- 
2.30.1

```

@michaelfuckner it would be nice if you could test this. If you would like a `Reported-by:` tag in the commit message, please give me an email address I can use. I will send this tomorrow if there are no additional comments.

**Comment by michaelfuckner** [comment]
looks good, please commit. please use michael@fuckner.net as mail

**Comment by nathanchance** [comment]
Patch submitted: https://lore.kernel.org/r/20210216213312.30462-1-nathan@kernel.org/

I am reopening this until it is merged into mainline.

**Comment by nathanchance** [comment]
Masahiro said he applied this but it has not hit linux-kbuild yet: https://lore.kernel.org/r/CAK7LNARcy52b6aGtV1mgq3rehtnwo7sm7TkNHc9H4bGu9Qdu8Q@mail.gmail.com/

I will post a SHA when that happens.

**Comment by nathanchance** [comment]
https://git.kernel.org/masahiroy/linux-kbuild/c/3d9055f99d240e42fd7c21e932ce33881fa81846

**Comment by michaelfuckner** [comment]
OK, I got a kernel compiled with aocc now, but building the Nvidia driver now needs "IGNORE_CC_MISMATCH=1", but it is also working. Thanks!

The major and minor number of the compiler used to
compile the kernel:

AMD clang version 11.0.0 (CLANG: AOCC_2.3.0-Build85 2020_11_10) (based on LLVM Mirror.Version.11.0.0), LLD 11.0.0 (/home/amd/jenkins/workspace/AOCC_2_3_0/llvm-project/lld bf979458a699d802d0b87f6c33d27ca1b00f4f53)

does not match the compiler used here:

AMD clang version 11.0.0 (CLANG: AOCC_2.3.0-Build#85 2020_11_10) (based on LLVM Mirror.Version.11.0.0)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/AMD/aocc-compiler-2.3.0/bin

**Comment by nathanchance** [comment]
Well, it does not match because of the fix for the initial report, right? We removed the `#` so that `make` did not error out, which makes the compiler text different by that one character. The true solution would be to report that to AMD and ask them to pick a different symbol to notate the build revision.

Anyways, this has been merged into mainline, so closing.

https://git.kernel.org/linus/c75173a26948363bdd11a0d5b90bd012ce4cc2e7
